package package kikaha.urouting.samples;

import io.undertow.server.HttpHandler;
import io.undertow.server.HttpServerExchange;
import javax.inject.Inject;
import javax.inject.Singleton;
import javax.enterprise.inject.Typed;
import kikaha.core.modules.http.WebResource;
import kikaha.urouting.RoutingMethodResponseWriter;
import kikaha.urouting.RoutingMethodParameterReader;
import kikaha.urouting.RoutingMethodExceptionHandler;
import kikaha.urouting.UndertowHelper;
import kikaha.urouting.api.Response;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Singleton
@Typed( HttpHandler.class )
@WebResource( path="/hello/world", method="GET" )
public class GeneratedRoutingMethod123 implements HttpHandler {

	@Inject RoutingMethodResponseWriter responseWriter;
	@Inject RoutingMethodParameterReader methodDataProvider;
	@Inject RoutingMethodExceptionHandler exceptionHandler;

	@Inject kikaha.urouting.samples.RetrieveRoutes provider;

	@Override
	public void handleRequest( HttpServerExchange exchange ) throws Exception {
        try {
			/* Does not require to parse form data. */
			/* Requires to parse body data. */
			exchange.getRequestReceiver()
				.receiveFullBytes((e, data) -> {
                    UndertowHelper.storeReadBodyData( e, data );
					runMethod( e );
				}, this::handleException );
            /* Does not require to parse body data. */
            runMethod( exchange );
        } catch ( Throwable cause ) {
            this.handleException( exchange, cause );
        }
	}

	private void runMethod( final HttpServerExchange exchange ){
		if ( exchange.isInIoThread() ) {
			exchange.dispatch( this::runMethod );
			return;
		}

		try {
				final String response = instance.renderRelatoMais( true );
					responseWriter.write( exchange, response );
		} catch ( Throwable cause ) {
			this.handleException( exchange, cause );
		}
	}

	private void handleException( final HttpServerExchange exchange, final Throwable cause ){
        if ( exchange.isInIoThread() ) {
            exchange.dispatch( ()-> handleException( exchange, cause ) );
            return;
        }

        try {
            final Response response = exceptionHandler.handle( cause );
            responseWriter.write( exchange, response );
        } catch ( Throwable newCause ){
            log.error( "Could not handle the failure." );
            exchange.endExchange();
        }
	}
}